feat: Add support for wrapping the devtools APIs

rpl opened the pull request:
This PR is applying the following changes to fix #56:

- the changes into api-metadata.json, needed to instruct the polyfill to wrap the devtools API namespace)
- a fix in the base browser-polyfill.js module to prevent the `TypeError` exception when the getter on the proxy is related to a non-configurable read-only property of the original Proxy target (fixed by using an empty object which has the original target as its prototype as the "root" target of the Proxy instance created by the polyfill)
- a fix in the base browser-polyfill.js module to force devtools.panels.create to return a promise that resolves to a single callback argument (instead of an resolving to an array of arguments which contains an additional undocumented second parameter)
- fix applied on one of the existent tests (related to the `Proxy TypeError` fix)
- 2 additional test case (which covers the `Proxy TypeError` fix and the new `maxResolvedArgs` property in the `devtools.panels.create` metadata).

rpl commented:
Hi @Rob--W 
do you mind to take a look at this PR?

Rob--W commented on src/browser-polyfill.js:
```diff
@@ -76,18 +76,30 @@ if (typeof browser === "undefined") {
+          if ("maxResolvedArgs" in metadata) {
+            // Some chrome APIs provides to their callback more arguments than expected.
+            callbackArgs = callbackArgs.slice(0, metadata.maxResolvedArgs);
+          }
```
There is no need to be so generic. Since the problem only occurs when with 1 vs not-one parameter, I suggest to throw away "maxResolvedArgs" and use `"singleCallbackParameter": true` or something similar. This could also be used to fix API definitions for callbacks that don't expect any parameter. Based on the code below, it is likely that callbacks without parameters resolve to an empty array.

Rob--W commented on test/test-async-functions.js:
```diff
@@ -86,5 +86,30 @@ describe("browser-polyfill", () => {
+          panels: {
+            create: sinon.spy((title, iconPath, panelURL, cb) => {
+              Promise.resolve().then(() => {
+                cb({isFakePanel: true}, "unwanted parameter");
```
Leave a comment what happens with `"unwanted parameter"`, and why.

Rob--W commented on test/test-proxied-properties.js:
```diff
@@ -75,28 +105,28 @@ describe("browser-polyfill", () => {
 
     it("deletes proxy getter/setter that are not wrapped", () => {
-      const fakeChrome = {};
+      const fakeChrome = {runtime: {}};
```
Why these changes? Please explain the changes in the commit description.

Rob--W commented:
I have filed an upstream bug for Chrome's misbehaving devtools.panels.create API: https://bugs.chromium.org/p/chromium/issues/detail?id=768159

rpl commented:
@Rob--W Thanks a lot for the review comments (and for filing the issue on Chromium issue tracker), I've applied the following changes based on you review comments:

- changed the new metadata property to `singleCallbackArg`, and added a test case related to a callback which expects 1 parameter called without any parameter (so that we test explicitly that the promise will resolve to a single undefined parameter instead of an empty array of callback parameters), and added to the test cases an inline comment that explains what behavior we expect

- squashed all the commits upfront and added a commit description comment which explains the changes applied (which included some additional information about the reason of the changes applied to the existent test case related to the "proxy setter/getter behaviors on non wrapped properties"): https://github.com/mozilla/webextension-polyfill/pull/57/commits/676966d0c28f5671d354de2c2882009db27e3ad4
tests

marcysutton commented:
Thank you so much for this! Excited to give it a go. Will report back in a new issue if I run into any problems.

marcysutton commented:
Do you know when this will get released? I'm not seeing it in 0.1.2. Thanks!

rpl commented:
@marcysutton I'm going to release a new version today (I'd like to include also #59, but if we are not able to land it today yet, I will release a new version for this change and I will include #59 in the next one).
